test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs

```diff
@@ -8958,6 +8958,95 @@ public void RestoreNetCore_PackagesLockFile_WithProjectChangeFramweorksAndLocked
+                   pathContext.SolutionRoot,
+                   NuGetFramework.Parse(projFramework));
+
+                var runtimeidentifiers = new List<string>(){ "win7-x64", "win-x86", "win", "z", "a" };
+                var ascending = runtimeidentifiers.OrderBy(i => i);
+                
+                projectA.Properties.Add("RuntimeIdentifiers", string.Join(";", ascending));
+
+                projectA.Properties.Add("RestorePackagesWithLockFile", "true");
+
+                // Set up the package and source
+                var packageX = new SimpleTestPackageContext()
+                {
+                    Id = "x",
+                    Version = "1.0.0"
+                };
+                packageX.Files.Clear();
+                packageX.AddFile("lib/netcoreapp2.0/x.dll");
+                packageX.AddFile("ref/netcoreapp2.0/x.dll");
+                packageX.AddFile("lib/net461/x.dll");
+                packageX.AddFile("ref/net461/x.dll");
+
+                await SimpleTestPackageUtility.CreateFolderFeedV3Async(
+                    pathContext.PackageSource,
+                    PackageSaveMode.Defaultv3,
+                    packageX);
+
+                projectA.AddPackageToAllFrameworks(packageX);
+
+
+                solution.Projects.Add(projectA);
+                solution.Create(pathContext.SolutionRoot);
+
+                // Act
+                var r = Util.RestoreSolution(pathContext);
+
+
+                Assert.True(File.Exists(projectA.NuGetLockFileOutputPath));
+                var packagesLockFile = PackagesLockFileFormat.Read(projectA.NuGetLockFileOutputPath);
+
+                //Modify the list of target/runtimes so they are reordered in the lock file
+                //Verify the passed in RIDs are within the lock file
+                //Verify the RIDS are not the same after reordering.
+                //Lock file is not ordered based on input RIDs so Validating the reorder here.
+                var originalTargets = packagesLockFile.Targets.Where(t => t.RuntimeIdentifier != null).Select(t => t.RuntimeIdentifier).ToList();
+                runtimeidentifiers.ShouldBeEquivalentTo(originalTargets);
+
+                packagesLockFile.Targets = packagesLockFile.Targets.
+                    OrderByDescending(t => t.RuntimeIdentifier==null).
+                    ThenByDescending( i => i.RuntimeIdentifier).ToList();
+                var reorderedTargets = packagesLockFile.Targets.Where(t => t.RuntimeIdentifier != null).Select(t => t.RuntimeIdentifier).ToList();
+
+                //The orders are not equal.  Then resave the lock file and project.
+                //The null RID must be the first one otherwise this fails
+                Assert.False(originalTargets.SequenceEqual(reorderedTargets));                
+                Assert.True(packagesLockFile.Targets[0].RuntimeIdentifier == null);
+                PackagesLockFileFormat.Write(projectA.NuGetLockFileOutputPath, packagesLockFile);
```

Comment:
Is the problem that the lock file is generated in a scrambled order(1), or that runtime identifiers are not sorted before comparing(2)? 

The test seems to be suggesting 1, but I thought it was 2.

Maybe the answer is 1, and 2 :)

Reply:
So what we are fixing with the PR is 2.  The RIDs you asked me to put in are not the order of targets coming out of the lock file.  The lock file comes out with a sorted alphabetical RIDs and frameworks.  The exception is the RID of null is on top and that is what contains all the dependencies.  So what I am doing in the test is 

- reading the targets from the lock file
- reordering them making sure the RID null node is on top
- Asserting that they are not equal before  re-saving.  This guarantees for us the new lock file we just saved, is out of order (compared to the one that was handed to us)
- We then write the new lock file and try the restore again.

Reply:
thanks for clarifying. 

Why do we need to reorder the lock file then?

Reply:
Because the [ProjectRuntimeKeys](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs#L105) is the thing changing order and we are comparing the lock file to it.  If it comes out via nuget its ordered, if its done in msbuild its not.

Reply:
oh you're forcing the rids to not be ordered because the test is nuget.exe. 

Now I get it, makes sense :) 

You can consider adding a dotnet restore test that wouldn't need the lock file reordered. 

https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetRestoreTests.cs#L261

Reply:
That might be the better path to go down.  The fact that we got 2 different results one from nuget.exe and one from dotnet, it might be worth having 2 tests that cover both utils.

Reply:
Will add a second test and ask for another once over.

Reply:
Yeah, I agree having both is better.

Reply:
Added, both tests fail without the fix, they both work with the fix.